xmodule/modulestore/django.py

```diff
@@ -365,34 +366,104 @@ class XBlockI18nService:
+            xblock_locale_domain, xblock_locale_dir = XBlockI18nService.get_python_locale_dir(block)
             selected_language = get_language()
-            try:
-                self.translator = gettext.translation(
-                    xblock_domain,
-                    xblock_locale_path,
-                    [to_locale(selected_language if selected_language else settings.LANGUAGE_CODE)]
-                )
-            except OSError:
-                # Fall back to the default Django translator if the XBlock translator is not found.
-                pass
+
+            if xblock_locale_dir:
+                try:
+                    self.translator = gettext.translation(
+                        xblock_locale_domain,
+                        xblock_locale_dir,
+                        [to_locale(selected_language if selected_language else settings.LANGUAGE_CODE)]
+                    )
+                except OSError:
+                    # Fall back to the default Django translator if the XBlock translator is not found.
+                    pass
+
+    @classmethod
+    def get_xblock_module_name(cls, block):
+        """
+        Return the XBlock Python module name.
+        """
+        xblock_class = getattr(block, 'unmixed_class', block.__class__)
+        xblock_module = xblock_class.__module__
+
+        if '.' in xblock_module:
+            xblock_module, _ = xblock_class.__module__.split('.')
+
+        return xblock_module
+
+    @classmethod
+    def get_python_locale_dir(cls, block):
+        """
+        Return the XBlock locale directory with the domain name.
+
+        Return:
+            (domain, locale_path): A tuple of the domain name and the XBlock locale directory.
+
+        This method looks for translations in two locations:
+            - First it looks for `atlas` translations in PLUGIN_TRANSLATIONS_ROOT.
+            - Alternatively, it looks for bundled translations in the XBlock pip package which are
+              found at <python_environment_xblock_root>/conf/locale/<language>/LC_MESSAGES/<domain>.po|mo
+
+        Note about the `text` and `django` domain:
+         - For unknown reasons, the `text` domain was used for XBlocks pre-OEP-58.
+         - As of OEP-58, the `django` domain is used for XBlocks
+           consistent with the other repositories.
+        """
+        xblock_module_name = cls.get_xblock_module_name(block)
+        xblock_locale_path = settings.REPO_ROOT / 'conf/plugins-locale/xblock.v1' / xblock_module_name / 'conf/locale'
+
+        # OEP-58 translations are pulled via atlas and takes precedence if exists.
+        if xblock_locale_path.isdir():
+            return 'django', xblock_locale_path
+
+        # Pre-OEP-58 translations within the XBlock pip packages are deprecated but supported.
+        deprecated_xblock_locale_path = resource_filename(xblock_module_name, 'translations')
+        return 'text', deprecated_xblock_locale_path
+
+    @classmethod
```

Comment:
Move the paths functions into a `xmodule/modulestore/api.py` module which can be imported by `common/xblock_django`.